fix: LEAP-897: Fix incorrect usage of name attr in visual tags

juliosgarbi opened the pull request:
### PR fulfills these requirements
- [x] Commit message(s) and PR title follows the format `[fix|feat|ci|chore|doc]: TICKET-ID: Short description of change made` ex. `fix: DEV-XXXX: Removed inconsistent code usage causing intermittent errors`
- [x] Tests for the changes have been added/updated (for bug fixes/features)
- [ ] Docs have been added/updated (for bug fixes/features)
- [x] Best efforts were made to ensure docs/code are concise and coherent (checked for spelling/grammatical errors, commented out code, debug logs etc.)
- [x] Self-reviewed and ran all changes on a local instance (for bug fixes/features)



#### Change has impacts in these area(s)
_(check all that apply)_
- [ ] Product design
- [ ] Backend (Database)
- [ ] Backend (API)
- [x] Frontend



### Describe the reason for change
This PR addresses a bug where annotations fail to save when users incorrectly assign a name attribute to visual tags that are also referenced by a Control tag. This issue was uncovered during scenarios where visual tags had name attributes and Control tags had toName assign to the name from the visual tag but exhibited no immediate error, leading to unsuccessful annotation saving.


#### What does this fix?
The application should validate the use of the name attribute on visual or experience tags and display an error if misused, to prevent incorrect configurations and ensure data integrity.

![Screenshot 2024-04-17 at 16 24 40](https://github.com/HumanSignal/label-studio/assets/4741623/09513b14-e396-4e6a-ad77-8df4905d7700)


#### Actual Behavior:
No error is displayed when a name attribute is added incorrectly, and annotations are not saved.


#### Impact:
This bug prevents users from completing annotations correctly, potentially causing data loss or corruption in projects relying on this functionality.


#### Solution:
Updated the tag validation logic to check for incorrect name attribute usage on visual tags, ensuring that an error is displayed to the user before saving annotations.


#### Additional Changes:
Included updates to configValidator unit tests to cover these scenarios and other scenarios that was broke.


### What level of testing was included in the change?
_(check all that apply)_
- [ ] e2e
- [ ] integration
- [x] unit

juliosgarbi committed:
fix: LEAP-897: Fix incorrect usage of name attr in visual tags

juliosgarbi committed:
fix unit tests for Config Validator

robot-ci-heartex committed:
ci: Build frontend

Workflow run: https://github.com/HumanSignal/label-studio/actions/runs/8727578742

Gondragos commented on web/libs/editor/src/core/DataValidator/ConfigValidator.js:
```diff
@@ -222,6 +222,21 @@ const validateParentTag = (element, model) => {
+ * @param {Object} element
+ */
+const validateVisualTags = (element) => {
+  const visualTags = ['Collapse', 'Filter', 'Header', 'Style', 'View'];
```
As potential improvement we may try to form this list from the information from `Registry`. But in that case we need new mixin like `BaseVisualTag` with volatile `isVisualTag: true` and mix it into all Visual tags. 
But it may be that I'm overthinking.

Gondragos replied:
Oh. We also may just use `model` as it is done in other validating functions. And check its `isVisualTag` when we add it to them.

hlomzik commented on web/libs/editor/src/core/__tests__/ConfigValidator.test.js:
```diff
@@ -4,42 +4,41 @@ import '../../tags/object/Image';
-// and this doesn't happen, may be that was some old behavior
-// @todo fix this
+function createStore(utterances) {
+  return { task: { dataObj: { utterances } } };
```
A little worrying that this task data is different from what actual configs in tests expect. Do they all fail just with data validation and not a config one?

juliosgarbi replied:
This validation is validating just the data, not the config. So doesn't matter which config we are using, if you write a test with everything good, it will return a empty array without error

hlomzik replied:
sorry, didn't get it... this validation specifically validates config. what am I missing?

juliosgarbi replied:
Yeah, I said everything wrong! Sorry, this validation is validating just the config, not the data

hlomzik commented on web/libs/editor/src/core/DataValidator/ConfigValidator.js:
```diff
@@ -301,6 +316,11 @@ export class ConfigValidator {
 
         if (parentValidation !== null) validationResult.push(parentValidation);
 
+        //validate visualTags
```
```suggestion
        // Validate visual tags
```

hlomzik commented on web/libs/editor/src/core/DataValidator/ConfigValidator.js:
```diff
@@ -222,6 +222,21 @@ const validateParentTag = (element, model) => {
+  const { tagName } = element;
+
+  if (visualTags.includes(tagName) && element.name) {
+    return errorBuilder.generalError('Attribute <b>name</b> is not allowed for tag <b>' + tagName + '</b>.');
```
```suggestion
    return errorBuilder.generalError(`Attribute <b>name</b> is not allowed for tag <b>${tagName}</b>.`);
```

juliosgarbi committed:
Apply suggestions from code review

Co-authored-by: hlomzik <hlomzik@gmail.com>

juliosgarbi committed:
remove configs from the tree to model

hlomzik commented:
/git merge
> [Workflow run](https://github.com/HumanSignal/label-studio/actions/runs/8755401572)
> Successfully merged:  45 files changed, 815 insertions(+), 790 deletions(-)

hlomzik committed:
Merge branch 'develop' into 'fb-leap-897'

Workflow run: https://github.com/HumanSignal/label-studio/actions/runs/8755401572

robot-ci-heartex committed:
ci: Build frontend

Workflow run: https://github.com/HumanSignal/label-studio/actions/runs/8755422492

juliosgarbi committed:
remove name from Header tags

juliosgarbi commented:
/git merge develop
> [Workflow run](https://github.com/HumanSignal/label-studio/actions/runs/8802810213)
> Successfully merged:  16 files changed, 847 insertions(+), 112 deletions(-)

juliosgarbi committed:
Merge branch 'develop' into 'fb-leap-897'

Workflow run: https://github.com/HumanSignal/label-studio/actions/runs/8802810213

robot-ci-heartex committed:
ci: Build frontend

Workflow run: https://github.com/HumanSignal/label-studio/actions/runs/8802839706

juliosgarbi commented:
/git merge develop
> [Workflow run](https://github.com/HumanSignal/label-studio/actions/runs/8818587422)
> Successfully merged:  6 files changed, 27 insertions(+), 9 deletions(-)

juliosgarbi committed:
Merge branch 'develop' into 'fb-leap-897'

Workflow run: https://github.com/HumanSignal/label-studio/actions/runs/8818587422

robot-ci-heartex committed:
ci: Build frontend

Workflow run: https://github.com/HumanSignal/label-studio/actions/runs/8818751278
